docs/proposals/koordlet/20230728-support-cold-memory-compute.md

```diff
@@ -0,0 +1,277 @@
+
+```go
+type KidledIdleInfo struct {
+	Version             string   `json:"version"`
+	PageScans           uint64   `json:"page_scans"`
+	SlabScans           uint64   `json:"slab_scans"`
+	ScanPeriodInSeconds uint64   `json:"scan_period_in_seconds"`
+	UseHierarchy        uint64   `json:"use_hierarchy"`
+	Buckets             []uint64 `json:"buckets"`
+	Csei                []uint64 `json:"csei"`
+	Dsei                []uint64 `json:"dsei"`
+	Cfei                []uint64 `json:"cfei"`
+	Dfei                []uint64 `json:"dfei"`
+	Csui                []uint64 `json:"csui"`
+	Dsui                []uint64 `json:"dsui"`
+	Cfui                []uint64 `json:"cfui"`
+	Dfui                []uint64 `json:"dfui"`
+	Csea                []uint64 `json:"csea"`
+	Dsea                []uint64 `json:"dsea"`
+	Cfea                []uint64 `json:"cfea"`
+	Dfea                []uint64 `json:"dfea"`
+	Csua                []uint64 `json:"csua"`
+	Dsua                []uint64 `json:"dsua"`
+	Cfua                []uint64 `json:"cfua"`
+	Dfua                []uint64 `json:"dfua"`
+	Slab                []uint64 `json:"slab"`
+}
+```
+
+Define a function named readidleInfo() in koordlet util module. It can read idleinfo memory.idle_stat which exists under every hierarchy of cgroup memory. (node e.g. /sys/fs/cgroup/memory/memory.idle_stat)
+
+```go
+func ReadIdleInfo(path string) (*IdleInfo, error) 
+```
+
+Define a function named GetIdleInfoFilePath(). It is used to return idleinfo file path. (node e.g. /sys/fs/cgroup/memory/memory.idle_stat)
+
+```go
+func GetIdleInfoFilePath(idleFileRelativePath string) string 
+```
+
+#### Report memory usage including hot page
+
+![image](../../images/support-cold-memory-2.svg)
+
+collectNodeMetric() is used to query node metric and return CPU And MemUsed in pkg/koordlet/statesinformer/impl/states_nodemetirc.go file.
+
+We can report memory usage including hot pages in collectNodeMetric().
+
+The calculation formulas of node, pod and container are as follows. 
+
+MemFree: free and unallocated memory
+NodeMemWithColdPage: cold page size in page cache of node
+
+node-memory-usage: nodeMemWithHotPageUasge=MemTotal-MemFree-NodeMemWithColdPage
+
+pod-memory-usage: podMemWithHotPageUasge=PodInactiveAnon + PodActiveAnon + PodUnevictable+PodMeMwithHotPageSize
+
+container-memory-usage: containerMemWithHotPageUasge=ContainerInactiveAnon + ContainerActiveAnon + ContainerUnevictable+ContainerMemWithHotPageSize
+
+The same process is executed pod informer to report memory usage. Do not repeat.
+
+#### Define cold memory API
+
+Provide memory-collect policy access.
+
+Add a field named MemoryCollectPolicy to represent which method is used to collect memory usage. Such as usageWithHotPageCache, usageWithoutPageCache,  usageWithPageCache.
```

Comment:
should clarify the definition when kilded is not installed if MemoryCollectPolicy == usageWithHotPageCache.
for example, report usageWithoutPageCache or nothing